### PR TITLE
Paper Trading capability

### LINKED ISSUE #135: Paper Trading capability
Paper trading is a great way to test out algorithms/bots without risking real money.  I've implemented a "Paper Exchange" and would be happy to share it with the community.

This capability also creates a stepping stone to support back-testing at a future point.

### PR DESCRIPTION
**Pull request by jleonelion**
this commit closes #135

### CODE REVIEW

**Comment by rmm5t** [comment]
There is very like a lot of great work here. However, there are several, if not a half dozen different pull requests wrapped up in this one PR.

Relevant Reading:
https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1

**Comment by jleonelion** [comment]
@rmm5t feel free to work with @DannyHinshaw and handle this however you'd like.  My prior PRs were rejected and I am including changes to satisfy the changes requested by the rejectors.  I never thought it would take so much effort to give away work.  All in all, I'm happy to share what I have done, but  not getting paid enough to justify the effort required to get these contributions into the mainline project.

**Comment by rmm5t** [comment]
Hi @jleonelion, I understand the frustrations. Just trying to help guide you towards an easier path of getting your hard work accepted (on any open source project). This is a good read as well:
https://www.igvita.com/2011/12/19/dont-push-your-pull-requests/

**Comment by DannyHinshaw** [comment]
@jleonelion I totally get the frustration. I really appreciate all the time and continued support for getting this feature working. I think we're on the home stretch, and as far as I know, once we address the order methods renaming mentioned by @blair (which I also asked about on [this most recent feature pull](https://github.com/jleonelion/gdax-tt/pull/2)) I should be able to finish this up for us, any maintenance related stuff to get this accepted etc... and speaking of...

@rmm5t thanks for the feedback. It would be very helpful to get more specific feedback from you as well. Not trying to skate by necessarily, but at this point it'd be great to know exactly what needs done to make this acceptable to get merged. The articles are good reads, but I wasn't able to extrapolate exactly what needs applied to this PR. I'm still sort of a pleb with git, but I'm willing to get my hands dirty, just need to know what needs done.

File: src/exchanges/paper/PaperExchange.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
 import { PlaceOrderMessage, isStreamMessage, TradeMessage, TradeExecutedMessage, TradeFinalizedMessage } from '../../core';
 import { BookBuilder, LiveOrder, AggregatedLevelWithOrders, Level3Order } from '../../lib';
-import { BigJS } from '../../lib/types';
+import { Big, BigJS } from '../../lib/types';
 import { AuthenticatedExchangeAPI, Balances } from '../AuthenticatedExchangeAPI';
 import { Product, PublicExchangeAPI, Ticker, CandleRequestOptions, Candle } from '../PublicExchangeAPI';
 import { HTTPError, GTTError } from '../../lib/errors';
-import { BigNumber } from 'bignumber.js';
+// import { BigNumber } from 'bignumber.js';
```
Comment (blair) [inline src/exchanges/paper/PaperExchange.ts]: Delete the commented out line.

File: src/exchanges/paper/PaperExchange.ts [hunk 2]
```diff
@@ -67,20 +68,20 @@ export class PaperExchange extends Duplex implements PublicExchangeAPI, Authenti
...
-            return Promise.reject(new GTTError('Order size must be a positive number'));
-        }
-        let orderPrice;
-        try {
-            orderPrice = new BigNumber(order.price);
-            if (orderPrice.lessThan(0)) { throw new Error('must be positive'); }
-        } catch (e) {
+
+        // util function, n must be a number (not undefined or NaN) and must be positive
+        const testOrderNum = (n: any): boolean => !n || isNaN(+n) || +n < 0;
```
Comment (blair) [inline src/exchanges/paper/PaperExchange.ts]: Suggest renaming testOrderNum to assertPositiveNumber() or something. You could also pass in the message number and have it throw the GTTError.

File: src/exchanges/paper/PaperExchange.ts [hunk 3]
```diff
@@ -247,7 +251,7 @@ export class PaperExchange extends Duplex implements PublicExchangeAPI, Authenti
         const orderBook = this.pendingOrdersByProduct.getValue(msg.productId);
         // do we have any pending orders for this product?
         if (orderBook !== undefined) {
-            const tradePrice = new BigNumber(msg.price);
+            const tradePrice = Big(msg.price) as BigJS;
```
Comment (blair) [inline src/exchanges/paper/PaperExchange.ts]: The 'as BigJS' shouldn't be necessary here.

**Comment by DannyHinshaw** [comment]
@blair thank for your feedback. I believe I have now addressed everything you requested. 

The latest commits aim to addresses the following:

• Revert the Trader class to it's previous states. The renaming of it's public methods for placing/cancelling orders would likely be unacceptable to push to the main branch (for anyone using those methods in the wild). This is solved this by using the executeMessage method directly when using the paper-trader.
• Small syntactic/sematic fixes requested by @blair
• Remove paper-exchange tests that remain pending when mocha has finished.

**Comment by DannyHinshaw** [comment]
@blair do you have permissions to review/merge this, or know who does?

**Comment by DannyHinshaw** [comment]
@fb55 could we get a review on this when you have the time?

File: README.md [hunk 4]
```diff
@@ -8,20 +8,104 @@
 Provide all the tools traders need, both professional and hobbyist alike, to create automated trading bots on the
 GDAX and supported digital asset exchanges. Note: Node 7.6 or above is required.
 
+## Getting Started
+### Install Prereqs
```
Comment (fb55) [inline README.md]: The readme does not need to be as detailed as this is. Especially the VS Code section is nice, but perhaps a bit much here.

File: src/core/Trader.ts [hunk 5]
```diff
@@ -18,13 +18,13 @@ import { AuthenticatedExchangeAPI } from '../exchanges/AuthenticatedExchangeAPI'
...
 import { Level3Order, LiveOrder, OrderbookState } from '../lib/Orderbook';
 import { CancelOrderRequestMessage,
-         isStreamMessage,
-         MyOrderPlacedMessage,
-         PlaceOrderMessage,
-         StreamMessage,
-         StreamMessageLike,
-         TradeExecutedMessage,
-         TradeFinalizedMessage } from './Messages';
+    isStreamMessage,
```
Comment (fb55) [inline src/core/Trader.ts]: Unnecessary whitespace changes

File: src/utils/Logger.ts [hunk 6]
```diff
@@ -17,7 +17,7 @@ import { sanitizeMessage } from '../core/Messages';
 export interface Logger {
     log(level: string, message: string, meta?: any): void;
 
-    error(err: Error): void;
+      error(err: Error): void;
```
Comment (fb55) [inline src/utils/Logger.ts]: This misalligns the method. I would prefer to see this file stay as it is.

**Comment by fb55** [comment]
Very hard to review this as it includes a lot of unrelated changes. I would very much prefer to have this be split into several PRs.

Also having some descriptions why certain changes were made would be nice. Eg. why do we include an S3 stream logger?

**Comment by fb55** [comment]
Generally I like the feature and would love to see it be part of the codebase, but the current state is not great for a review.

**Comment by DannyHinshaw** [comment]
@fb55 thank you, what can I do to make it better?

File: src/core/Trader.ts (hunk 5 above)
Comment (DannyHinshaw) [reply src/core/Trader.ts]: This has been reverted, not sure why it's still showing in changed files...

**Comment by DannyHinshaw** [comment]
@fb55 please re-review at your leisure. I have manually gone through and removed anything extraneous or unrelated to the paper-trader feature specifically.

**Comment by DannyHinshaw** [comment]
@fb55 , please close this PR without merging. I'm going to create a feature branch instead of rewriting over all of jleon's changes on master, rebase and work out a few more kinks. I'll submit a new PR when I'm finished and you can re-review then.

**Comment by fb55** [comment]
Sounds good — thanks a lot for looking into this!
